Add integration for Swift-MesonLSP

JCWasmx86 opened the pull request:
See #119

### Features
- [x] Download and setup binaries for MacOS/Windows (For windows: Requires a zip file due to depending on some DLLs that can't be statically linked)
  - [x] Download binaries
  - [x] Check checksum
  - [x] Extract zip file (Requires external library)
  - [x] Make lsp.ts use the downloaded file
- [x] Code style/Formatting
- [x] Figure out if something went wrong with yarn.lock and if yes what went wrong (Doesn't seem to be something bad)
- [x] Get rid of curl usage to download file
- [x] Don't require users to restart VSCode 
- [ ] Add language server to https://langserver.org/ ? (See https://github.com/JCWasmx86/Swift-MesonLSP/issues/11)

JCWasmx86 commented:
I think this is at least partially ready for review (Sadly I wasn't able to find a way - even with help of others - to download a file that worked for me. It would always end up empty)

Automatic updates are currently not implemented as I don't know what the best way would be.

This adds a few dependencies:
- [which](https://www.npmjs.com/package/which) to find the language server on the path
- [adm-zip](https://www.npmjs.com/package/adm-zip) to extract the downloaded zip

lsp.ts is based a bit on the [Vala VSCode extension (MIT)](https://github.com/vala-lang/vala-vscode)

tristan957 commented on package.json:
```diff
@@ -177,6 +177,16 @@
           },
           "description": "Give an object that's merged into the debug configuration provider"
+        },
+        "mesonbuild.languageServerPath": {
```
It is possible that there could be multiple language server implementations in the future. Perhaps these settings could take inspiration from the Python extension because I think it supports multiple?

JCWasmx86 replied:
That's implicitly supported. You just have to point it to an executable that `--lsp --stdio` as parameters. Sure not optimal

tristan957 replied:
I think this extension having a blessed default (yours) makes the most sense. Then someone could change the setting if another language server ever cropped up.

I don't think any pop up is necessary other than asking do you want to download the latest version of the configured language server.

JCWasmx86 replied:
Had an idea: Maybe instead of downloading, setup everything automatically, if a supported distribution is found. That would require root rights, but I'm sure this could work in some way

JCWasmx86 replied:
This thread will be obsoleted by https://github.com/JCWasmx86/vscode-meson/pull/3

tristan957 commented on package.json:
```diff
@@ -177,6 +177,16 @@
+          "description": "Binary name or path to language server",
+          "default": ""
+        },
+        "mesonbuild.neverDownloadLanguageServer": {
```
I would say drop the never here

tristan957 commented on package.json:
```diff
@@ -323,7 +333,13 @@
     "@types/node": "^16.11.7",
     "@types/vscode": "^1.1.59",
-    "typescript": "^4.4.4"
+    "typescript": "^4.4.4",
```
nit pick, but maybe alphabetize.

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,77 @@
+  ls: LanguageClient | null = null
+
+  constructor(_context: ExtensionContext) {
+    this.config = workspace.getConfiguration('vala');
```
Is 'vala' right here?

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,77 @@
+
+  dispose() {
+    if (this.ls) {
+      this.ls!.stop()
```
Is the exclamation mark really necessary here?

tristan957 commented:
The clangd, codeql, rust analyzer extensions might be good to look at for inspiration on auto updating.

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  startLanguageServer(serverModule: string): void {
+    let clientOptions: LanguageClientOptions = {
```
Seems like there could be more "const" instead of "let" here, but I am definitely no expert :).

tristan957 commented on src/utils.ts:
```diff
@@ -149,6 +154,135 @@ export function extensionConfigurationSet<
+export function canDownloadLanguageServer(): boolean {
+  const platform = os.platform()
+  if (platform != "win32" && platform != "darwin") {
+    return false
```
Style: Inconsistency when if statements are only one line.

tristan957 commented on src/utils.ts:
```diff
@@ -149,6 +154,135 @@ export function extensionConfigurationSet<
+}
+
+function createHashForLanguageServer(): string {
+  const platform = os.platform()
```
A switch would probably be better here and other places, but could be wrong.

tristan957 commented on src/utils.ts:
```diff
@@ -149,6 +154,135 @@ export function extensionConfigurationSet<
+  }
+}
+
+
```
extra new line

tristan957 commented on src/utils.ts:
```diff
@@ -149,6 +154,135 @@ export function extensionConfigurationSet<
+  )
+  const tmpPath = path.join(os.tmpdir(), `lsp-${Date.now()}.zip`);
+  try {
+    const promise = downloadFile(createDownloadURL(), tmpPath);
```
Any reason not to just use await here?

tristan957 commented on src/utils.ts:
```diff
@@ -149,6 +154,135 @@ export function extensionConfigurationSet<
+}
+
+const getExtensionDir = (): string => {
+  return path.join(os.homedir(), ".vscode-meson");
```
Doesn't VSCode give extensions a place to store data?

JCWasmx86 replied:
I'm sure there is one, but I didn't find it :)

tristan957 replied:
https://code.visualstudio.com/api/extension-capabilities/common-capabilities#data-storage

JCWasmx86 replied:
Yes, I found it already ^^ Nevertheless thanks

tristan957 commented on src/utils.ts:
```diff
@@ -149,6 +154,135 @@ export function extensionConfigurationSet<
+}
+
+function createDownloadURL(): string {
+  const platform = os.platform()
```
I think some of these functions should be getters that are only computed the first time. Perhaps creating a `MesonLanguageSever` abstract class, and then you create an implementation for your client. Your client would obviously be the only implementation.

```typescript
abstract class MesonLanguageServer {
  name: string;
  serverOptions: ServerOptions;

  constructor() {
    // This would be your current setupLanguageServer maybe?
  }

  get downloadInfo(): [url: string, hash: string] | null {
    return null;
  }

  private dispose(): void {

  }

  restart(): void {
    // ...
  }
}
```

Hopefully you get what I am saying

JCWasmx86 commented on package.json:
```diff
@@ -177,6 +177,21 @@
+            "never",
+            "yes"
+          ],
+          "description": "Have VSCode download the language server automatically (MacOS/Windows only)"
```
I'm no expert, but you have to consider e.g. NixOS with its weird fixed up paths, different glibc versions etc. (All the reasons flatpak exists)

That's why my/dcbaker's idea (No sure from whom this idea was😅) was to just use distro packages. For Arch I setup an AUR, for Fedora a COPR. dcbaker wanted to write a nix expression. For Ubuntu/Debian I wanted to setup an APT repo sometime.

JCWasmx86 replied:
Did search a bit more: Linking with musl is currently (22 Jun. 2022) not possible. So using the package managers is the best action

JCWasmx86 commented on src/lsp.ts:
```diff
@@ -0,0 +1,77 @@
+      transport: TransportKind.stdio
+    };
+
+    this.ls = new LanguageClient("Swift-MesonLSP", "Meson Language Server", serverOptions, clientOptions, true);
```
Tbh I like the name, it says everything "language server for meson" and "written in Swift" or "It's fast", depending if you know Swift ^^

dcbaker commented:
I'm sorry I've been really behind on this, lots of RL stuff right now. I will try to get a nix expression written so we can test the distro provided mechanism

JCWasmx86 commented:
No worries. I've already implemented that it searches in this order:

1. The extension internal directory (Where the language server was downloaded)
2. The executable from the config
3. path

JCWasmx86 commented:
For updates I found two ways:

### Using Github releases
Just download the latest artifacts from Github.
### Maintain separate JSON
E.g. have some JSON file
```
{
  "versions": {
    "2.0": {
      "win32-x64": {
        "hash": "1122334455",
        "url": "https://someurl"
      },
    }
  }
}
```
The keys could be a concatenation of os.platform() and os.arch(), thus dynamically allowing other os/arch combinations.

dcbaker commented:
There's some rebase issues that will need to be sorted, but the code itself looks good to me. I'm going to write a nix expression and give this a test drive. I forgot to look, did you add an entry for this to the `CHANGELOG`?

JCWasmx86 commented:
I can resolve the conflicts later. I've not added a changelog section yet.

JCWasmx86 commented:
I've resolved the conflicts and added an section to the changelog including an optional very concise help for Linux users

tristan957 commented on CHANGELOG.md:
```diff
@@ -2,6 +2,9 @@
 
+- Add integration for [Swift-MesonLSP](https://github.com/JCWasmx86/Swift-MesonLSP).
+  vscode-meson will automatically download it on Windows and MacOS. Linux users should
+  follow the [docs](https://github.com/JCWasmx86/Swift-MesonLSP/tree/main/Docs)
```
Add a newline.

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,59 @@
+    setupLanguageServer(this)
+  }
+
+  dispose() {
```
It has been a while since I wrote TS. Do you have to mark this private?

JCWasmx86 replied:
Not sure, but making it private breaks nothing

tristan957 replied:
I only say that because it is only called in finalize.

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  get languageServerPath(): string | null {
+    return findLanguageServer(this) || this.config.languageServerPath || which.sync(this.executableName, { nothrow: true })
```
The `this.config.languageServerPath` should probably come first right?

tristan957 commented on src/utils.ts:
```diff
@@ -123,6 +128,118 @@ export function extensionConfigurationSet<
   return getConfiguration().update(key, value, target);
 }
 
+export function canDownloadLanguageServer(client: LanguageServerClient): boolean {
```
Can't you use LanguageServerClient::donwloadInfo() instead of having this function? Perhaps it just got left by accident

tristan957 commented:
Now the CI can run since I have the relevant permissions :)

tristan957 commented on src/utils.ts:
```diff
@@ -123,6 +128,110 @@ export function extensionConfigurationSet<
+export function findLanguageServer(client: LanguageServerClient): string | null {
+  const platform = os.platform();
+  const executableName = client.executableName
+  if (platform != "win32" && platform != "darwin")
```
Some of this logic seems specific to your language server. Might try to find a way to move this into the abstract class somehow?

JCWasmx86 replied:
Maybe some computed properties that returns all supported OS/Arch combinations?

tristan957 replied:
Perhaps. I am thinking of a `LanguageServerClient::find()` function or some sort. You could mark the function as not abstract, and then use the computed property like you said. Might not need each subclass to write their own impl that way.

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  async setupLanguageServer() {
+    setupLanguageServer(this)
```
Seems like you should just inline the other function to me. No need to call an out of band function.

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,58 @@
+    return this.config.languageServerPath || findLanguageServer(this) || which.sync(this.executableName, { nothrow: true })
+  }
+
+  async setupLanguageServer() {
```
I would mark the return type as `Promise<void>`

JCWasmx86 commented:
utils.ts is no unchanged again compared to main, but broke something somewhere as mkdirSync fails now with ENOENT

JCWasmx86 commented:
Should work again. Mixed up path of an URL and the fsPath

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,160 @@
+    });
+  }
+
+  findLanguageServer(): string | null {
```
I think this can be made private?

tristan957 commented on src/lsp/swift-mesonlsp.ts:
```diff
@@ -0,0 +1,64 @@
+    return os == "win32" || os == "darwin";
+  }
+
+  startLanguageServer(serverModule: string): void {
```
I think this function can be a private function to the abstract parent class. The `clientOptions` should probably be a private member of that base class. I don't see it changing for other potential clients. Do you?

I think `runExe` and `debugExe` could probably be abstract getters on the base class? Then `serverOptions` could be made using the abstract getters.

Overall, I think this design is coming together quite nicely.

JCWasmx86 replied:
debugExe/runExe depend on `serverModule`, so best I can do is a method that e.g. returns an array of executables. Or we could mandate, that all implementations have to accept `--lsp`?

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,160 @@
+    const tmpPath = path.join(os.tmpdir(), `lsp-${Date.now()}.zip`);
+    try {
+      this.downloadFile(downloadInfo[0], tmpPath).then(() => {
+        const hashPromise = this.computeFileHash(tmpPath);
```
what do you think about moving the hash check to the download function?

JCWasmx86 replied:
Not a fan of it, as it won't be composable anymore. One function = one functionality

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,160 @@
+    )
+    const tmpPath = path.join(os.tmpdir(), `lsp-${Date.now()}.zip`);
+    try {
+      this.downloadFile(downloadInfo[0], tmpPath).then(() => {
```
Should we clean up the file if anything of the following code `reject`s

JCWasmx86 replied:
Not a fan of it, as it won't be composable anymore. One function = one functionality

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,160 @@
+    this.startLanguageServer(serverModule);
+  }
+
+  abstract startLanguageServer(serverModule: string): void
```
Trailing semicolons!

tristan957 commented on src/lsp/swift-mesonlsp.ts:
```diff
@@ -0,0 +1,64 @@
+import { LanguageServerClient } from "../lsp";
+
+export class SwiftMesonLspLanguageClient extends LanguageServerClient {
+  ls: LanguageClient | null = null
```
I don't think is necessary? Looks like you already have one in the parent class.

tristan957 commented on src/lsp/swift-mesonlsp.ts:
```diff
@@ -0,0 +1,64 @@
+            "c10999081cef8b56ea164d9a170470c268045254829e48bf4ac137a07a8a874c"]
+  }
+
+  supportsSystem(os: string, arch: string): boolean {
```
Doesn't your language server support linux?

JCWasmx86 replied:
Yes but no  setup here in the extension due to the issues discussed elsewhere

JCWasmx86 replied:
I have two ideas regarding this:
- Automatically setup the package manager on supported OS. This requires elevation to root (
- Upload more binaries for different distributions to Github Releases (Compiles on Ubuntu 18.04/20.04/22.04, Fedora, Arch). But it would be outside of the package manager and at least I would be annoyed it some extension downloads some binary albeit Linux has a package manager.

JCWasmx86 commented:
On my TODO list is still:  "Should muon and Swift-MesonLSP be XOR? Swift-MesonLSP uses muon for formatting and duplicates a lot of work from there. So you would have e.g. duplicate diagnostics"

Should I simply disable the muon integration, if a running language server is detected?

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,188 @@
+
+    if (serverModule == null)
+      return;
+    this.startLanguageServer(serverModule);
```
Why do you need to pass `serverModule` here? The function you are classing has access to the `this.languageServerPath`.

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,188 @@
+  private startLanguageServer(serverModule: string) {
+    const runExe: Executable = {
+      command: serverModule,
+      args: ["--lsp"],
```
After you fix the previous comment do the abstract getter for exposing the run and debug exes. The subclasses can access this.languageServerPath and not need to take serverModule as an argument anymore

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,188 @@
+
+  abstract get downloadInfo(): [url: string, hash: string] | null;
+
+  abstract get repoURL(): string;
```
I don't think this has to be a get?

It can be

```
abstract repoURL: string;
```

tristan957 commented on src/extension.ts:
```diff
@@ -189,10 +190,44 @@ export async function activate(ctx: vscode.ExtensionContext) {
+  const languageServerKey = "languageServer"
+  const lsKey = extensionConfiguration(languageServerKey);
+  const possibleServers = {
+    "Swift-MesonLSP": new SwiftMesonLspLanguageClient(ctx)
```
Don't preconstruct the clients. Wasted resources.

```
const client = possibleServers[key](ctx);

JCWasmx86 commented:
Would moving the calls
```
activateLinters(root, ctx);
activateFormatters(ctx);
```
in extension.ts either below the question whether you want to install a language server/if a language server is installed or should I move my code over these two lines?

tristan957 commented:
Just pick one and run with it imo.

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,180 @@
+import { Executable, LanguageClient, LanguageClientOptions, ServerOptions, TransportKind } from "vscode-languageclient/node";
+
+export abstract class LanguageServerClient {
+  config: WorkspaceConfiguration;
+  ls: LanguageClient | null = null;
+  executableName: string;
+  context: ExtensionContext;
+  private clientOptions: LanguageClientOptions = {
+    documentSelector: ["meson", { "scheme": "file", language: "meson" }]
+  };
+  abstract repoURL: string;
+  abstract get debugExe(): Executable
+  abstract get runExe(): Executable
```
Some of these should probably be marked as private/protected. Also ending semicolons

JCWasmx86 replied:
Will do that, do you have any linter/formatter for this repo that can do that automatically?

tristan957 replied:
@lukester1975 has a PR to add prettier support which I think is a good idea. Then we can also work on adding eslint and various other tools as needed to aid in the development/review process.

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,180 @@
+    this.executableName = executableName;
+    this.context = context;
+    this.config = workspace.getConfiguration("mesonbuild");
+    const serverModule = this.languageServerPath;
```
get rid of this temp variable

tristan957 replied:
I see now why you were passing in serverModule into startLanguageServer. Hmm

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,180 @@
+      transport: TransportKind.stdio
+    };
+
+    this.ls = new LanguageClient("Swift-MesonLSP", "Meson Language Server", serverOptions, this.clientOptions, true);
```
This needs to be made more generic :)

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,180 @@
+    this.ls.start();
+  }
+
+  canDownloadLanguageServer(): boolean {
```
I think the logic in this function is strange. Whether you can download the language server is different that if you have currently found it on the current system

JCWasmx86 replied:
I have difficulties finding a name. That's the workflow I've thought of (Pseudo code):


1. If the language server is installed, just start it and do nothingmore
2. If the OS is linux and a supported distribution+version: Download script that sets up the package sources and installs it
3. If the OS is linux and is not a supported distribution: Redirect user to the browser with the instructions to install from source
4. If the OS is Windows/MacOS: Just download+extract the zip files

A fallback could be just taking a nearly statically linked Ubuntu 18.04 binary and hoping for the best

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,180 @@
+  }
+
+  get languageServerPath(): string | null {
+    return this.config.languageServerPath || this.findLanguageServer() || which.sync(this.executableName, { nothrow: true })
```
Is it possible to just resolve the language server path in the constructor?

JCWasmx86 replied:
No, as e.g. the PATH can change (Drop binary into /usr/bin) and findLanguageServer() returns the path to the downloaded executable => Not cachable if I understand correctly

tristan957 replied:
Ok, I think we should just resolve the path in the constructor, and add a command to "restart the language server" similar to clangd. Then a user can change the path of the language server all they want.

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,180 @@
+    return path.join(ctx.globalStorageUri.fsPath, "lsp");
+  }
+
+  public dispose() {
```
this should be made private

JCWasmx86 replied:
I got this error:
```
src/extension.ts:292:26 - error TS2345: Argument of type 'SwiftMesonLspLanguageClient' is not assignable to parameter of type '{ dispose(): any; }'.
  Property 'dispose' is private in type 'SwiftMesonLspLanguageClient' but not in type '{ dispose(): any; }'.

292   ctx.subscriptions.push(client)
                             ~~~~~~


Found 1 error in src/extension.ts:292
```
After changing to private. Just having `dispose () {` seems to compile.

tristan957 replied:
Oh, I guess it is required by VSCode. Nevermind me :)

tristan957 commented on src/lsp/swift-mesonlsp.ts:
```diff
@@ -0,0 +1,47 @@
+export class SwiftMesonLspLanguageClient extends LanguageServerClient {
+  repoURL: string = "https://github.com/JCWasmx86/Swift-MesonLSP";
+  get runExe(): Executable {
+    return {
+    command: this.languageServerPath,
+    args: ["--lsp"]
+    }
+  }
+  get debugExe(): Executable {
+    return {
+    command: this.languageServerPath,
+    args: ["--lsp"]
```
formatting

tristan957 commented on src/lsp/swift-mesonlsp.ts:
```diff
@@ -0,0 +1,47 @@
+    if (arch != "x64")
+      return null;
+    if (platform == "win32")
+      return ["https://github.com/JCWasmx86/Swift-MesonLSP/releases/download/v2.0/Swift-MesonLSP-win64.zip",
```
```
const artifacts: { [key: string]: { name: string, hash: string } } = {
  "win32-x64": { name: "Swift-MesonLSP-win64.zip", hash: "fc0d930eb67525309291907c07df8655920cdc651f17f2320d09cc5271a23876" },
  "darwin-x64": { name: "Swift-MesonLSP-macos12.zip", hash: "c10999081cef8b56ea164d9a170470c268045254829e48bf4ac137a07a8a874c" },
};

const artifact = artifacts[`${platform}-${arch}`];
return [`${this.repoURL}/releases/download/v2.0/${artifact.name}`, artifact.hash];
```

tristan957 replied:
`artifacts` can be made a private member variable.

tristan957 commented on src/lsp/swift-mesonlsp.ts:
```diff
@@ -0,0 +1,47 @@
+
+  get downloadInfo(): [url: string, hash: string] | null {
+    const platform = os.platform();
+    if (platform != "win32" && platform != "darwin")
+      return null;
+    const arch = os.arch();
+    if (arch != "x64")
+      return null;
```
There is duplicated logic here and in supportsSystem

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,180 @@
+      transport: TransportKind.stdio
+    };
+
+    this.ls = new LanguageClient("Swift-MesonLSP", "Meson Language Server", serverOptions, this.clientOptions, true);
```
What is the second argument to this constructor?

JCWasmx86 replied:
It's this string here:

![image](https://user-images.githubusercontent.com/63594136/233178083-b8de0051-2aad-40bb-b051-dec48efb2008.png)

tristan957 replied:
Oh. Should the whole Meson extension use the output channel? To me that would make sense. What are your thoughts?

tristan957 replied:
But maybe that is not possible? If not, then this is fine :).

JCWasmx86 replied:
There is e.g. extra output from my language server that would e.g. help to diagnose issues and some other stuff

![image](https://user-images.githubusercontent.com/63594136/233179506-ff9e6878-e127-4805-9f45-06df53631d1e.png)


![image](https://user-images.githubusercontent.com/63594136/233179555-9632a6fa-d327-4e56-bb19-11fcf96bed6b.png)

tristan957 replied:
Is there any way to just add a prefix to log messages coming from your LS?

JCWasmx86 replied:
I don't know, as I'm not that well-versed with the VSCode API

tristan957 replied:
Just leave as is, and we can explore this more later since it really isn't that important.

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,180 @@
+    return null;
+  }
+
+  private getExtensionLSPDir(ctx: vscode.ExtensionContext): string {
```
We don't need to take ctx here since you are storing it.

tristan957 commented:
> Disabling muon when your language server is detected sounds good to me.

Logging a message to the output channel of the extension would be good in this case too.

tristan957 commented on src/lsp.ts:
```diff
@@ -0,0 +1,180 @@
+    try {
+      this.downloadFile(downloadInfo[0], tmpPath).then(() => {
+        this.computeFileHash(tmpPath).then(str => {
+          const expected = downloadInfo[1];
```
I think when you have named tuples, you can use `downloadInfo.hash`

JCWasmx86 replied:
No, does not seem to work:

```

> mesonbuild@1.9.0 compile
> tsc -p ./

src/lsp/index.ts:84:41 - error TS2339: Property 'hash' does not exist on type '[url: string, hash: string]'.

84           const expected = downloadInfo.hash;
                                           ~~~~


Found 1 error in src/lsp/index.ts:84
```

tristan957 replied:
Interesting. Had no idea. I guess it is purely for documentation purposes then.

tristan957 commented on src/lsp/swift-mesonlsp.ts:
```diff
@@ -0,0 +1,47 @@
+import { Executable } from "vscode-languageclient/node";
+
+export class SwiftMesonLspLanguageClient extends LanguageServerClient {
+  private artifacts: { [key: string]: { name: string, hash: string } } = {
```
static just for good measure.

tristan957 commented:
Move the src/lsp.tx file to src/lsp/index.ts

tristan957 commented:
I now have your branch pulled locally.

tristan957 commented:
Might as well rebase on your next push. I think the only actual comment left at the moment is adding a "restart language server command". Then, it seems like we should restart the language server if "mesonbuild.languageServer" OR "mesonbuild.languageServerPath" change.

tristan957 commented on src/extension.ts:
```diff
@@ -199,10 +198,54 @@ export async function activate(ctx: vscode.ExtensionContext) {
+    activateLinters(root, ctx);
+    activateFormatters(ctx);
+  } else {
+    getOutputChannel().appendLine("Not enabling muon linter/formatter as language server is active!");
```
```suggestion
    getOutputChannel().appendLine("Not enabling muon linter/formatter as language server is active!");
    getOutputChannel().appendLine("Not enabling the muon linter/formatter because the the language server is active!");
```

tristan957 commented on package.json:
```diff
@@ -86,6 +86,10 @@
         "title": "Meson: Run Benchmarks"
+      },
+      {
+        "command": "mesonbuild.restart",
```
Maybe the command name should be "mesonbuild.restartLanguageServer"

tristan957 commented on src/extension.ts:
```diff
@@ -248,4 +291,12 @@ export async function activate(ctx: vscode.ExtensionContext) {
     }
   }
+
+  ctx.subscriptions.push(client);
+
+  ctx.subscriptions.push(
+    vscode.commands.registerCommand("mesonbuild.restart", async () => {
+      client.restart();
+    })
+  );
```
What is `client` when a binary isn't found?

JCWasmx86 replied:
If no binary is found, it basically does nothing.

tristan957 commented on src/lsp/index.ts:
```diff
@@ -0,0 +1,180 @@
+
+  abstract get downloadInfo(): [url: string, hash: string] | null;
+
+  restart(): void {
```
In this function, we want to re-resolve the language server's path.

JCWasmx86 replied:
It is doing this automatically, as it  is computed again by referencing `languageServerPath`, if I understand correctly

tristan957 replied:
You are correct on this aspect, but I think it would be best to resolve the path in the constructor as mentioned in a previous comment.

JCWasmx86 replied:
I don't understand what should be resolved in the constructor. Resolving `which.sync(this.executableName, { nothrow: true })` doesn't make sense, as there is a workflow like:

- Linux user starts VSCode
- Installs LS into /usr/local/bin manually
- Restarts the language server

And because this part is resolved earlier, the language server would still fail to change.

`this.findLanguageServer()` doesn't make sense too:
- Windows/MacOS user starts VSCode
- LS is installed
- `restart()`
- The already resolved result from `findLanguageServer()` is still null

And for `this.config.languageServerPath` I don't see a use case, too.
- User changes config, switches to some other language server
- `restart()`
- The already resolved value would still be the old language server

So what exactly do you mean?

JCWasmx86 commented:
I've added a temporary commit that shows another approach. A language server can be setup in two ways: Automatically and manually. For Windows/MacOS it is automatically, while Linux users are redirected to the documentation.

While I've not given up on the idea of basically doing `curl https://github.com/some/setup/script/for/linux.sh | sudo sh` to automatically setup the package sources for integration with the systems' package manager, I think there would be some hurdles (Would it work in confined environments like flatpak/snaps etc.)

So requiring manual setup for Linux users is probably inevitable unless you want to either download a binary for an ancient glibc (E.g. Ubuntu 18.04) and hope to the best. But AFAIK it would make problems for e.g. NixOS.  Or you use a script that will probably fail on some systems due to sandboxing or other issues.

tristan957 commented on package.json:
```diff
@@ -86,6 +86,10 @@
+      },
+      {
+        "command": "mesonbuild.restartLanguageServer",
+        "title": "Meson: Restart language server"
```
Title case. "Meson: Restart Language Server"

tristan957 commented on src/extension.ts:
```diff
@@ -199,12 +199,81 @@ export async function activate(ctx: vscode.ExtensionContext) {
 
-    if (configureOnOpen === true) {
-      runFirstTask("reconfigure");
+  const downloadLanguageServerKey = "downloadLanguageServer";
```
Can you create an enum for the settings instead of having these strings located throughout the code?

JCWasmx86 replied:
Do you mean like this: https://github.com/mesonbuild/vscode-meson/pull/123/commits/dbb15f11b248687894976efa299ba23904775c84#diff-c54113cf61ec99691748a3890bfbeb00e10efb3f0a76f03a0fd9ec49072e410aR133

tristan957 commented on src/lsp/index.ts:
```diff
@@ -0,0 +1,183 @@
+  private ls: LanguageClient | null = null;
+  executableName: string;
+  private context: ExtensionContext;
+  private clientOptions: LanguageClientOptions = {
```
This should be private static

tristan957 commented on src/lsp/index.ts:
```diff
@@ -0,0 +1,183 @@
+export abstract class LanguageServerClient {
+  private config: WorkspaceConfiguration;
+  private ls: LanguageClient | null = null;
+  executableName: string;
```
Does this need to be abstract?

tristan957 commented:
I have been hacking on your branch. I will submit a PR to your remote.

JCWasmx86 commented:
I'm not sure how I should restructure the commits, as there are e.g. temporary commits contained within and some by @tristan957. I would just make it one commit `Add integration for Swift-MesonLSP` and manually update the authorship

tristan957 commented:
Squash it all down assuming you're happy. Just add me with a coauthor trailer.

tristan957 commented:
I'll take another look at the PR in the next couple of days.

JCWasmx86 commented:
Oops, messed up the commit history of prior commits

JCWasmx86 commented:
I'm available again for addressing reviews. Could you please review especially the last commit? `tmp: Controversial changes.`. 

Either it's my somewhat mediocre TS+async model skills, or there was something wrong with the code before (As I probably did some mistake during testing).


- With the hashing I had problems due to `reject`ing for some unknown reason. I fixed it by moving everything into the returned promised and removing the `pipe` to the hash-object.
- I switched from a promise-based architecture to a "normal" architecture, as I couldn't get changes to the local variable `uri` to be returned. Furthermore the code is more clear now, ymmv.

JCWasmx86 commented:
Thinking more about this, there should be some way to disable the language server for a project. There are at least three projects that use meson that would get *a lot* of errors due to the project structure:

- [HelenOS](http://www.helenos.org/)
- [picolibc](https://github.com/picolibc/picolibc)
- [DPDK](https://www.dpdk.org/)

All of these stem from the fact that I haven't implemented an interpreter, instead I'm basically guessing the parameters to `set_variable` and `subdir` using some [weird mix of heuristics and partially interpreting](https://github.com/JCWasmx86/Swift-MesonLSP/blob/main/Sources/MesonAnalyze/SetVariableGuess.swift)

HelenOS is basically not possible for my language server, as it has a config system that generates parts of the build AFAIU. 

Both picolibc and DPDK have a code structure that is more difficult for my partial interpreter to understand

tristan957 commented:
That makes sense to me. Wanna add an option for it?

JCWasmx86 commented:
Can do that

tristan957 commented on CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
 ## next
 
+- Add integration for [Swift-MesonLSP](https://github.com/JCWasmx86/Swift-MesonLSP).
```
While you are rebasing, s/for/with

tristan957 commented on package.json:
```diff
@@ -187,6 +191,35 @@
+          "default": "Swift-MesonLSP",
+          "enum": [
+            "Swift-MesonLSP",
+            "null"
```
Can we use the null keyword in JSON instead here?

tristan957 commented on package.json:
```diff
@@ -187,6 +191,35 @@
+          "enumDescriptions": [
+            "Download the language server",
+            "Do not download the language server",
+            "Ask every time"
```
Might be worth finishing this string with `to download the language server`

tristan957 commented on src/extension.ts:
```diff
@@ -186,25 +184,80 @@ export async function activate(ctx: vscode.ExtensionContext) {
 
         case Options.never:
-          extensionConfigurationSet(configureOnOpenKey, false, vscode.ConfigurationTarget.Workspace);
+          extensionConfigurationSet(SettingsKey.ConfigureOnOpen, false, vscode.ConfigurationTarget.Workspace);
```
I would like the casing of `ConfigureOnOpen` to match `configureOnOpen`. Similar for the other keys

tristan957 commented on src/extension.ts:
```diff
@@ -186,25 +184,80 @@ export async function activate(ctx: vscode.ExtensionContext) {
+    return false;
   }
 
+  let client = await createLanguageServerClient(server, await shouldDownload(downloadLanguageServer), ctx);
```
It looks like this could be const.

JCWasmx86 replied:
No, as a few lines later I do:
```ts
  ctx.subscriptions.push(
    vscode.commands.registerCommand("mesonbuild.restartLanguageServer", async () => {
      if (client === null) {
        client = await createLanguageServerClient(server, await shouldDownload(downloadLanguageServer), ctx); // <<- HERE
        if (client !== null) {
          ctx.subscriptions.push(client);
          client.start();
        }
      } else {
        client.restart();
      }
    })
  );
```

tristan957 commented on src/extension.ts:
```diff
@@ -186,25 +184,80 @@ export async function activate(ctx: vscode.ExtensionContext) {
+    ctx.subscriptions.push(client);
+    client.start();
+
+    getOutputChannel().appendLine("Not enabling the muon linter/formatter because the language server is active.");
```
Only output this line if the language server is yours (granted it is the only one haha). Also explicitly add Swift-MesonLsp to the output line.

tristan957 commented on src/extension.ts:
```diff
@@ -186,25 +184,80 @@ export async function activate(ctx: vscode.ExtensionContext) {
+        client = await createLanguageServerClient(server, await shouldDownload(downloadLanguageServer), ctx);
+        if (client !== null) {
+          ctx.subscriptions.push(client);
+          client.start();
```
The output line from above about not enabling muon would be good to have here, though I am not sure of the best factoring.

JCWasmx86 replied:
Wouldn't we have to stop linting using muon there? And as far as I can see, there's currently no way to do that

tristan957 replied:
Hmm good point. Wanna add a TODO on that front for if we ever figure it out?

JCWasmx86 replied:
Yes, will do that

tristan957 commented on src/lsp/common.ts:
```diff
@@ -0,0 +1,53 @@
+import { SwiftMesonLspLanguageClient } from "./swift-mesonlsp";
+import { Uri } from "vscode";
+
+export async function createLanguageServerClient(server: LanguageServer, download: boolean, context: vscode.ExtensionContext): Promise<LanguageServerClient | null> {
```
I think this might make sense to factor into the LanguageServerClient class.

```typescript
public static async LanguageServerClient.create(...): Promise<LanguageServerClient | null>;
```

JCWasmx86 replied:
Sadly moving it does not work:
```
2023-08-08 18:45:28.158 [error] TypeError: Class extends value undefined is not a constructor or null
	at Object.<anonymous> (/home/user/Projects/vscode-meson/out/lsp/swift-mesonlsp.js:6:49)
	at Module._compile (node:internal/modules/cjs/loader:1141:14)
	at Module._extensions..js (node:internal/modules/cjs/loader:1196:10)
	at Module.load (node:internal/modules/cjs/loader:1011:32)
	at Module._load (node:internal/modules/cjs/loader:846:12)
	at f._load (node:electron/js2c/asar_bundle:2:13330)
	at o._load (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:130:34104)
	at f._load (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:130:30479)
	at I._load (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:94:24270)
	at Module.require (node:internal/modules/cjs/loader:1035:19)
	at require (node:internal/modules/cjs/helpers:102:18)
	at Object.<anonymous> (/home/user/Projects/vscode-meson/out/lsp/index.js:15:26)
	at Module._compile (node:internal/modules/cjs/loader:1141:14)
	at Module._extensions..js (node:internal/modules/cjs/loader:1196:10)
	at Module.load (node:internal/modules/cjs/loader:1011:32)
	at Module._load (node:internal/modules/cjs/loader:846:12)
	at f._load (node:electron/js2c/asar_bundle:2:13330)
	at o._load (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:130:34104)
	at f._load (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:130:30479)
	at I._load (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:94:24270)
	at Module.require (node:internal/modules/cjs/loader:1035:19)
	at require (node:internal/modules/cjs/helpers:102:18)
	at Object.<anonymous> (/home/user/Projects/vscode-meson/out/lsp/common.js:5:12)
	at Module._compile (node:internal/modules/cjs/loader:1141:14)
	at Module._extensions..js (node:internal/modules/cjs/loader:1196:10)
	at Module.load (node:internal/modules/cjs/loader:1011:32)
	at Module._load (node:internal/modules/cjs/loader:846:12)
	at f._load (node:electron/js2c/asar_bundle:2:13330)
	at o._load (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:130:34104)
	at f._load (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:130:30479)
	at I._load (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:94:24270)
	at Module.require (node:internal/modules/cjs/loader:1035:19)
	at require (node:internal/modules/cjs/helpers:102:18)
	at Object.<anonymous> (/home/user/Projects/vscode-meson/out/extension.js:13:18)
	at Module._compile (node:internal/modules/cjs/loader:1141:14)
	at Module._extensions..js (node:internal/modules/cjs/loader:1196:10)
	at Module.load (node:internal/modules/cjs/loader:1011:32)
	at Module._load (node:internal/modules/cjs/loader:846:12)
	at f._load (node:electron/js2c/asar_bundle:2:13330)
	at o._load (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:130:34104)
	at f._load (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:130:30479)
	at I._load (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:94:24270)
	at Module.require (node:internal/modules/cjs/loader:1035:19)
	at require (node:internal/modules/cjs/helpers:102:18)
	at Function.i [as __$__nodeRequire] (/usr/share/codium/resources/app/out/vs/loader.js:5:98)
	at r._loadCommonJSModule (/usr/share/codium/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:130:32106)
	at async Promise.all (index 0)
```
That was the reason in the beginning I moved it into the common.ts. The explanation according to SO was some cyclic import/dependency cycle or something like that.

tristan957 replied:
I'll pull the latest branch this week to try it for myself just to see what I can dig up

tristan957 commented on src/lsp/swift-mesonlsp.ts:
```diff
@@ -0,0 +1,64 @@
+
+export class SwiftMesonLspLanguageClient extends LanguageServerClient {
+  private static artifacts: { [key: string]: { name: string, hash: string } } = {
+    "win32-x64": { name: "Swift-MesonLSP-win64.zip", hash: "1bb06cec242251bcf7ea58e173a46105039fbdf381c3c0ab6edd3b7c3798dbfa" },
+    "darwin-x64": { name: "Swift-MesonLSP-macos12.zip", hash: "e016c9041f57ed452ff8d45fb2a28e82dd446d7606bf18107b572b3cbd82fd20" },
+    "linux-x64": { name: "Swift-MesonLSP.zip", hash: "0a8d1eecccbc35d4b3d49444824a44c5da2430a34692f706ba34acaf12322efa" }
```
If you have had any new releases, might as well update the refs

tristan957 reviewed (commented):
Just these last few comments!

JCWasmx86 commented:
For some reason the download/hashing/extracting is horrendously broken. (I thought I fixed it a while ago, but not sure why it exactly fails)

It seems to be easy:

- Download
- Hash
- Extract

But for some reason it does:
- Download + Hash concurrently, so the hash ends up being an empty string => Fail

Have to debug that a bit more.

tristan957 commented:
Sounds good. I am on Linux with your COPR enabled, so I should be able to test the cyclic import thing on my end :)

JCWasmx86 commented:
Sadly I still wasn't able to figure out, why it attempts to hash an empty string instead of the contents, maybe you have a better luck with finding the cause for it

JCWasmx86 commented:
I got it working now. For some reason github returned 302 instead of 300/301, which I wasn't able to handle => This PR is ready for maybe the final review?

tristan957 commented:
Cool wanna squash the commits where appropriate. I wanna try to get to it this upcoming week. Thanks for your continued work.

JCWasmx86 commented:
```
Run tsc --noEmit --noErrorTruncation --pretty false
Error: src/utils.ts(31,33): error TS2322: Type 'ExecFileException | undefined' is not assignable to type 'ExecException | undefined'.
  Type 'ExecFileException' is not assignable to type 'ExecException | undefined'.
    Type 'ExecFileException' is not assignable to type 'ExecException'.
      Types of property 'code' are incompatible.
        Type 'string | number | null | undefined' is not assignable to type 'number | undefined'.
          Type 'null' is not assignable to type 'number | undefined'.
Error: Process completed with exit code 2.
```
Not sure why exactly it breaks, but I broke it

tristan957 commented:
Did you rebase on master?

JCWasmx86 commented:
Yes, already reformatted everything

JCWasmx86 commented:
I've accidentally bumped the `@types/node` version during fixing conflicts, but it's fixed now

tristan957 commented:
Ha. Thanks for rebasing :).

tristan957 commented on CHANGELOG.md:
```diff
@@ -7,6 +7,11 @@
 - Add syntax highlight for `meson.options` files.
 - Fix listing tests in the Test Explorer on activation
 - Add support to run executables that are not installed (eg: examples)
+- Add integration with
+  [Swift-MesonLSP](https://github.com/JCWasmx86/Swift-MesonLSP). vscode-meson
+  will automatically download it on Windows, Linux and MacOS. Linux users can
+  follow the [docs](https://github.com/JCWasmx86/Swift-MesonLSP/tree/main/Docs)
+  if they want to have updates delivered using their package manager.
```
Move this to the next section

tristan957 commented on package.json:
```diff
@@ -338,14 +371,20 @@
-  "dependencies": {},
+  "dependencies": {
+    "adm-zip": "0.5.10",
+    "vscode-languageclient": "8.0.2",
```
Mind updating vscode-languageclient? Let's add `^` to these versions
